[docs] Gh #5196 describe current limitations (#5428)
DevExpress/testcafe: docs/articles/documentation/guides/basic-guides/select-page-elements.md

</html>
```

> If a selector matches multiple elements, the subsequent methods return results for all the elements. The following selector returns the child nodes of all `<div>` tags on the page, uncluding their children and so on:

Reviewer: > uncluding

Author: edited

> If a selector matches multiple elements, the subsequent methods return results for all the elements. The following selector returns the child nodes of all `<div>` tags on the page, including their children and so on:

Reviewer 2: ```suggestion
If a selector matches multiple elements, the subsequent methods return results for all elements. The following selector returns the child nodes of all `<div>` tags on the page, including their children:
```

Author: Doesn't this imply that only parents and their immediate children are affected? Because this actually affects their children and their children and their children until we reach the root.

Reviewer 2: I see now. How about "including their children and their decendants"?

Or maybe simply "including their decendants"?
Let me know what you think.

Author: It thinks this works well:

> If a selector matches multiple elements, the subsequent methods return results for all elements. The following selector returns the child nodes of all `<div>` tags on the page, including their children and their descendants:

</html>
```

> If a selector matches multiple elements, only the first matching element will be used for [actions] or when retrieving [DOM Node Snapshots](../../reference/test-api/domnodestate.md). Use [`.nth`](../../reference/test-api/selector/nth.md) method to enumerate through other matching elements.

Reviewer: > `[actions]`

This was supposed to be a link?

Author: edited

> If a selector matches multiple elements, only the first matching element will be used for [actions](./interact-with-the-page.md) or when retrieving [DOM Node state](../../reference/test-api/domnodestate.md). Use [.nth](../../reference/test-api/selector/nth.md) method to enumerate through other matching elements.

Reviewer 2: `Use [.nth](../../reference/test-api/selector/nth.md) method to enumerate through other matching elements.`

"enumerate through" is not correct. Are you using "enumerate" in the context of "count" or "list". Or do you actually mean "iterate through"?

Author: Yes, iterate is the word.

Reviewer 2: ```suggestion
If a selector matches multiple elements, only the first matching element will be used for [actions](./interact-with-the-page.md) or when retrieving the [DOM Node state](../../reference/test-api/domnodestate.md). Use the [.nth](../../reference/test-api/selector/nth.md) method to iterate through other matching elements.
```

> You can retrieve the number of matched elements with a [selector.count](../../reference/test-api/selector/count.md) property.

Reviewer 2: ```suggestion
You can use the [selector.count](../../reference/test-api/selector/count.md) property to retrieve the number of matched elements.
```

});
```

#### DOM Node State

> To use an object's state multiple times in the test, get the object that contains [all the data](../../reference/test-api/domnodestate.md) in one turnaround to the client. To obtain this object (*DOM Node State*), call the selector with the `await` keyword:

Reviewer 2: ```suggestion
To use an object's state multiple times in the test, get the object that contains [all data](../../reference/test-api/domnodestate.md) in one turnaround to the client. To obtain this object (*DOM Node State*), call the selector with the `await` keyword:
```

Use this approach for Node.js callbacks that fire during the test run. To ensure that the test function
does not finish before the callback is executed, suspend the test until the callback fires. For instance, you can introduce a Promise and wait until it completes synchronously, as shown in the example above.

> The `boundTestRun` option requires the same test controller instance that is passed to the function used in a test declaration. It's unable to work with imported test controllers.
>

> > The `boundTestRun` option is unable to pass a test controller instance to a function declared in another module or class.

Reviewer: `boundTestRun` is so 'unable' :smile:
Maybe we can change at least one occurrence to 'does not accept' or something?

Reviewer: And I'm not sure what this second note means

Author: I removed the redundancy

Use this approach for Node.js callbacks that fire during the test run. To ensure that the test function
does not finish before the callback is executed, suspend the test until the callback fires. For instance, you can introduce a Promise and wait until it completes synchronously, as shown in the example above.

> > The `boundTestRun` option requires the same test controller instance that is passed to the function used in a test declaration. It's unable to work with imported test controllers.

Reviewer 2: ```suggestion
> The `boundTestRun` option requires the same test controller instance that is passed to the function used in the test declaration. It cannot work with imported test controllers.
```

});
```

### Select Elements That Contain Special Characters

> If your page contains [HTML symbols](https://www.w3schools.com/html/html_symbols.asp) or [HTML entities](https://www.w3schools.com/html/html_entities.asp) (like `&nbsp;`, newline chars), use their [unicode counterparts](https://www.rapidtables.com/code/text/unicode-characters.md) in [Selector.WithText](../../reference/test-api/selector/withtext.md) and [Selector.WithExactText](../../reference/test-api/selector/withexacttext.md).

Reviewer 2: ```suggestion
If your page contains [HTML symbols](https://www.w3schools.com/html/html_symbols.asp) or [HTML entities](https://www.w3schools.com/html/html_entities.asp) (e.g., `&nbsp;`, newline chars), use their [unicode counterparts](https://www.rapidtables.com/code/text/unicode-characters.md) in [Selector.WithText](../../reference/test-api/selector/withtext.md) and [Selector.WithExactText](../../reference/test-api/selector/withexacttext.md).
```

});
```

### Select Elements Containing Special Characters

If your page contains special HTML characters, also known as [HTML entities](https://www.w3schools.com/html/html_entities.asp) (like `&nbsp;`, newline chars), use their [unicode counterparts](https://www.rapidtables.com/code/text/unicode-characters.md) in [`Selector.WithText`](../../reference/test-api/selector/withtext.md) and [`Selector.WithExactText`](../../reference/test-api/selector/withexacttext.md).

**Example**

```html
<html>
    <body>
        <p>Click&nbsp;me</p>
    </body>
</html>
```

```js
import { Selector } from 'testcafe';

fixture `My fixture`
    .page `http://localhost/`;

test('My test', async t => {

>     const sel = await Selector('p').withText('Click&nbsp;me') //typed representation, does not work

Reviewer: I think it's safer to say

`//typed representation, not supported`

Author: edited

    const sel = await Selector('p').withText('Click\u00a0me') //unicode representation, works
    const sel = await Selector('p').withText('Click\xa0me') //hexadecimal representation, works

>     const sel = await Selector('p').withText('Click\160me') //decimal representation introduced with an octal escape sequence, throws an error

Reviewer: maybe we can mention the strict mode in the comment? otherwise, it seems like we have an issue here:

```
    const sel = await Selector('p').withText('Click\160me') // decimal representation introduced with an octal escape sequence;
                                                            // not supported because tests are executed in strict mode
```

Author: edited
